src/compose/app.ts

```diff
@@ -658,10 +660,33 @@ export class App {
 		);
 	}
 
-	private generateContainerStep(current: Service, target: Service) {
-		// if the services release doesn't match, then rename the container...
+	private generateContainerStep(
+		current: Service,
+		target: Service,
+		locksTaken: LocksTakenMap,
+	) {
+		// Update container metadata if service release has changed
 		if (current.commit !== target.commit) {
-			return generateStep('updateMetadata', { current, target });
+			// Only update service once locked or if legacy container
+			const shouldSkipLock = checkTruthy(
+				current.config.labels['io.balena.legacy-container'],
+			);
+			// QUESTION: Should updateMetadata only be allowed when
+			// *all* services have locks taken by the Supervisor? Currently
+			// it proceeds when the service it's updating has locks taken,
+			// meaning the service could be on new release while another service
+			// with a user-taken lock is still on old release.
+			if (
+				shouldSkipLock ||
+				locksTaken.isLocked(target.appId, target.serviceName)
+			) {
+				return generateStep('updateMetadata', { current, target });
+			}
+			// Otherwise, take lock for service first
+			return generateStep('takeLock', {
```

Comment:
Could we take locks for all services after downloads are finished? Feels strange taking locks one by one since basically all locks will need to be taken: changing services, services that require updateMetadata and new services

Reply:
> Could we take locks for all services after downloads are finished? 

I don't think it's that simple, what about an app where update strategies differ between services? Locks need to be taken before kill, so the lock would need to be taken for the entire app before downloads finish if a service has `kill-then-download`, leaving the lock in place unnecessarily for services with `download-then-kill` which may still be early in their download progress

Reply:
Actually, I think https://github.com/balena-os/balena-supervisor/commit/e836028efcb284827010b0d38f99242bd9127063 already makes it so that a lock is taken for all services before a kill if `kill-then-download`. This could be undesirable for services with `download-then-kill` which are then stuck with the lock even when they haven't finished downloading yet

Reply:
If there are any services with `delete-then-download` or `kill-then-download` then all locks should be taken before executing any further steps.

Reply:
Okay, since the code already does that, we should be good to not modify any more logic around update strategies